File: lib/java/opentoken/src/test/java/com/truveta/opentoken/attributes/validation/NotStartsWithValidatorTest.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.attributes.validation;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.util.Collections;
import java.util.Set;

import org.junit.jupiter.api.Test;

class NotStartsWithValidatorTest {

    @Test
    void testNullValue() {
        NotStartsWithValidator validator = new NotStartsWithValidator(Collections.singleton("invalid"));
        assertFalse(validator.eval(null), "Null value should not be allowed");
    }

    @Test
    void testEmptyInvalidPrefixes() {
        NotStartsWithValidator validator = new NotStartsWithValidator(Collections.emptySet());
        assertTrue(validator.eval("anyvalue"), "Empty prefix sets should allow any value");
    }

    @Test
    void testValueNotStartingWithInvalidPrefix() {
        NotStartsWithValidator validator = new NotStartsWithValidator(Collections.singleton("bad"));
        assertTrue(validator.eval("goodvalue"), "Values not starting with invalid prefixes should be allowed");
    }

    @Test
    void testValueStartingWithInvalidPrefix() {
        NotStartsWithValidator validator = new NotStartsWithValidator(Collections.singleton("bad"));
        assertFalse(validator.eval("badvalue"), "Values starting with invalid prefixes should not be allowed");
    }

    @Test
    void testExactMatch() {
        NotStartsWithValidator validator = new NotStartsWithValidator(Collections.singleton("bad"));
        assertFalse(validator.eval("bad"), "Values that exactly match invalid prefixes should not be allowed");
    }

    @Test
    void testMultipleInvalidPrefixes() {
        NotStartsWithValidator validator = new NotStartsWithValidator(Set.of("bad", "evil", "wrong"));
        assertTrue(validator.eval("goodvalue"), "Valid values should be allowed");
        assertFalse(validator.eval("badvalue"), "Values starting with 'bad' should not be allowed");
        assertFalse(validator.eval("evilplan"), "Values starting with 'evil' should not be allowed");
        assertFalse(validator.eval("wrongway"), "Values starting with 'wrong' should not be allowed");
        assertTrue(validator.eval("rightway"), "Values not starting with any invalid prefix should be allowed");
    }

    @Test
    void testCaseSensitive() {
        NotStartsWithValidator validator = new NotStartsWithValidator(Collections.singleton("Bad"));
        assertTrue(validator.eval("badvalue"), "Validation should be case-sensitive");
        assertFalse(validator.eval("Badvalue"), "Values starting with exact case match should not be allowed");
    }

    @Test
    void testEmptyStringValue() {
        NotStartsWithValidator validator = new NotStartsWithValidator(Collections.singleton("bad"));
        assertTrue(validator.eval(""), "Empty string should be allowed if it doesn't start with invalid prefix");
    }

    @Test
    void testEmptyStringPrefix() {
        NotStartsWithValidator validator = new NotStartsWithValidator(Collections.singleton(""));
        assertFalse(validator.eval("anyvalue"), "Any value should be invalid if empty string is an invalid prefix");
        assertFalse(validator.eval(""), "Empty string should be invalid if empty string is an invalid prefix");
    }

    @Test
    void testSerialization() throws Exception {
        NotStartsWithValidator validator = new NotStartsWithValidator(Set.of("bad", "evil"));
        ByteArrayOutputStream byteOut = new ByteArrayOutputStream();
        ObjectOutputStream out = new ObjectOutputStream(byteOut);
        out.writeObject(validator);

        ByteArrayInputStream byteIn = new ByteArrayInputStream(byteOut.toByteArray());
        ObjectInputStream in = new ObjectInputStream(byteIn);
        NotStartsWithValidator deserializedValidator = (NotStartsWithValidator) in.readObject();

        // Test that both validators behave the same way
        String[] testValues = { "goodvalue", "badvalue", "evilplan", "rightway", "", null };

        for (String value : testValues) {
            assertEquals(validator.eval(value), deserializedValidator.eval(value),
                    "Validators should behave identically for value: " + value);
        }

        // Additional specific tests on deserialized validator
        assertTrue(deserializedValidator.eval("goodvalue"));
        assertFalse(deserializedValidator.eval("badvalue"));
        assertFalse(deserializedValidator.eval("evilplan"));
    }
}

File: lib/python/opentoken/src/test/opentoken/attributes/validation/not_starts_with_validator_test.py
"""
Tests for NotStartsWithValidator.
"""

import pickle
import pytest

from opentoken.attributes.validation.not_starts_with_validator import NotStartsWithValidator


class TestNotStartsWithValidator:
    """Test cases for NotStartsWithValidator."""

    def test_null_value(self):
        """Test that null (None) values are not allowed."""
        validator = NotStartsWithValidator({"invalid"})
        assert validator.eval(None) is False, "Null value should not be allowed"

    def test_empty_invalid_prefixes(self):
        """Test that an empty set allows any value."""
        validator = NotStartsWithValidator(set())
        assert validator.eval("anyvalue") is True, "Empty prefix sets should allow any value"

    def test_value_not_starting_with_invalid_prefix(self):
        """Test that values not starting with invalid prefixes are allowed."""
        validator = NotStartsWithValidator({"bad"})
        assert validator.eval("goodvalue") is True, "Values not starting with invalid prefixes should be allowed"

    def test_value_starting_with_invalid_prefix(self):
        """Test that values starting with invalid prefixes are not allowed."""
        validator = NotStartsWithValidator({"bad"})
        assert validator.eval("badvalue") is False, "Values starting with invalid prefixes should not be allowed"

    def test_exact_match(self):
        """Test that values exactly matching invalid prefixes are not allowed."""
        validator = NotStartsWithValidator({"bad"})
        assert validator.eval("bad") is False, "Values that exactly match invalid prefixes should not be allowed"

    def test_multiple_invalid_prefixes(self):
        """Test that multiple invalid prefixes are handled correctly."""
        validator = NotStartsWithValidator({"bad", "evil", "wrong"})
        assert validator.eval("goodvalue") is True, "Valid values should be allowed"
        assert validator.eval("badvalue") is False, "Values starting with 'bad' should not be allowed"
        assert validator.eval("evilplan") is False, "Values starting with 'evil' should not be allowed"
        assert validator.eval("wrongway") is False, "Values starting with 'wrong' should not be allowed"
        assert validator.eval("rightway") is True, "Values not starting with any invalid prefix should be allowed"

    def test_case_sensitive(self):
        """Test that validation is case-sensitive."""
        validator = NotStartsWithValidator({"Bad"})
        assert validator.eval("badvalue") is True, "Validation should be case-sensitive"
        assert validator.eval("Badvalue") is False, "Values starting with exact case match should not be allowed"

    def test_empty_string_value(self):
        """Test that empty string values are allowed if they don't start with invalid prefixes."""
        validator = NotStartsWithValidator({"bad"})
        assert validator.eval("") is True, "Empty string should be allowed if it doesn't start with invalid prefix"

    def test_empty_string_prefix(self):
        """Test that empty string as a prefix invalidates all values."""
        validator = NotStartsWithValidator({""})
        assert validator.eval("anyvalue") is False, "Any value should be invalid if empty string is an invalid prefix"
        assert validator.eval("") is False, "Empty string should be invalid if empty string is an invalid prefix"

    def test_serialization(self):
        """Test serialization and deserialization of the validator."""
        validator = NotStartsWithValidator({"bad", "evil"})

        # Serialize the validator using pickle
        serialized_data = pickle.dumps(validator)

        # Deserialize the validator
        deserialized_validator = pickle.loads(serialized_data)

        # Test that both validators behave the same way
        test_values = ["goodvalue", "badvalue", "evilplan", "rightway", "", None]

        for value in test_values:
            original_result = validator.eval(value)
            deserialized_result = deserialized_validator.eval(value)

            assert original_result == deserialized_result, (
                f"Validators should behave identically for value: {value} "
                f"(original: {original_result}, deserialized: {deserialized_result})"
            )

        # Additional specific tests on deserialized validator
        assert deserialized_validator.eval("goodvalue") is True
        assert deserialized_validator.eval("badvalue") is False
        assert deserialized_validator.eval("evilplan") is False